python/numpy/weldarray.py

```diff
@@ -0,0 +1,375 @@
+                output = outputs[0]
+            else:
+                # not sure about the use case with multiple outputs - let numpy
+                # handle it.
+                supported = False
+                output = None
+        else:
+            output = None
+
+        return output, supported
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        '''
+        Overwrites the action of weld supported functions and for others,
+        pass on to numpy's implementation.
+
+        FIXME: Not dealing with method at all. Check use cases.
+        TODO: add input descriptions.
+        '''
+        input_args = [inp for inp in inputs]
+        outputs = kwargs.pop('out', None)
+        output, supported = self._process_ufunc_inputs(input_args, outputs)
+        if output is None:
+            # the new array will have no views parent/children
+            self.view_parents = []
+            self.view_children = []
+
+        # check for supported ops.
+        if ufunc.__name__ in UNARY_OPS and supported:
+            assert(len(input_args) == 1)
+            return self._unary_op(UNARY_OPS[ufunc.__name__], result=output)
+
+        if ufunc.__name__ in BINARY_OPS and supported:
+            # self can be first or second arg.
+            if isinstance(input_args[0], WeldArray):
+                # first arg is WeldArray, must be self
+                assert input_args[0].name == self.name
+                other_arg = input_args[1]
+            else:
+                other_arg = input_args[0]
+                assert input_args[1].name == self.name
+
+            return self._binary_op(other_arg, BINARY_OPS[ufunc.__name__],
+                    result=output)
+
+        # Relegating the work to numpy. If input arg is WeldArray, evaluate it,
+        # and convert to ndarray before passing to super()
+        for i, arg_ in enumerate(input_args):
+            if isinstance(arg_, WeldArray):
+                # Evaluate all the lazily stored computations first.
+                input_args[i] = arg_._eval()
+        if outputs:
+            out_args = []
+            for j, output in enumerate(outputs):
+                if isinstance(output, WeldArray):
+                    out_args.append(output.view(np.ndarray))
+                else:
+                    out_args.append(output)
+            kwargs['out'] = tuple(out_args)
+        else:
+            outputs = (None,) * ufunc.nout
+
+        result = super(WeldArray, self).__array_ufunc__(ufunc, method,
+                                                 *input_args, **kwargs)
+        # if possible, return weldarray.
+        if str(result.dtype) in SUPPORTED_DTYPES:
+            return WeldArray(result)
+        else:
+            return result
+
+    def evaluate(self):
+        '''
+        User facing function - if he wants to explicitly evaluate all the
+        registered ops.
+        '''
+        return WeldArray(self._eval())
+
+    def _eval(self):
```

Comment:
actually it does do caching I think - every time _eval is called, and there are registered ops, it evaluates them, and creates a new weldobject with no ops registered, so next time self.name == weldobj.weld_code is true because of the way I'd set it up. maybe I should make this more explicit?